Compact ui

sam-ao opened the pull request:
Created a compact version of the telemetry GUI. Much of the back end source code was stripped as it would no longer be used by the compact UI.

sam-ao committed:
finished front end compact ui

sam-ao committed:
completed backend for compactui

sam-ao committed:
did some cleanup of unused files

enochtsang commented:
@JordanHeinrichs Please look through the pull request, I'd like us both to look through it before merging it in.

JordanHeinrichs commented on SolarCarTelemetry.sublime-workspace:
```diff
@@ -0,0 +1,216 @@
+{
```
I think this file was accidentally added. If you want you can add a gitignore for *.sublime-workspace.

JordanHeinrichs commented on ViewLayer/CompactUI/CompactUI.h:
```diff
@@ -0,0 +1,193 @@
+    QLabel& setOutput3GConnectionStatus();
+
+private:
+    double mppt1CurrentIn;
```
So far within this project all private class members have an underscore following the name. I think it would be a good thing to keep consistent.

JordanHeinrichs commented on ViewLayer/CompactUI/CompactUI.cpp:
```diff
@@ -0,0 +1,582 @@
+CompactUI::CompactUI()
+    :ui(new Ui::CompactUI)
+{
+    mppt1CurrentIn=0;
```
I am still trying to understand the full scope of the changes so I might retract this. But generally a UI class should not double up as an entity. My first instinct would be that if you need to store these values within the data layer.

JordanHeinrichs commented:
I did a first quick overview. I want to pull down the repo and try it out tomorrow. I will finish my feedback before Saturday.

sam-ao committed:
fixed some warnings

sam-ao committed:
fixed a minor error

sam-ao committed:
restructured MpptData

JordanHeinrichs committed:
Fixed build.

Removed sublime project files.

JordanHeinrichs commented:
With regards to the compactUi fault pane, personally when I was driving the car I have never looked at the fault history dialog but just at the red/green indicators. I would be for having an abbreviated version of that instead of the fault history. 

Since there is limited space for faults I would choose these faults:

Motor:
software Overcurrent
hardware Overcurrent

Battery:
Cell over voltage
Cell under voltage
Cell over temperature
CMU communications timeout.


Cell under voltage
Cell over temperature
CMU communications timeout.

For the motor ones make sure it is motor 1 fault status || motor 2 fault status. Although if there is enough space it would be nice to have separate indicators for each motor.

JordanHeinrichs commented on ViewLayer/CompactView/CompactView.cpp:
```diff
@@ -0,0 +1,826 @@
+#include "../../PresenterLayer/PowerPresenter/PowerPresenter.h"
+#include "../../PresenterLayer/VehiclePresenter/VehiclePresenter.h"
+#include "../../PresenterLayer/MpptPresenter/MpptPresenter.h"
+#include "../../PresenterLayer/FaultsPresenter/FaultsPresenter.h"
```
For future reference, you can just have `#include "PresenterLayer/FaultsPresenter/FaultsPresenter.h"` the file paths can just be from the include directory and not relative from the file.

sam-ao committed:
restructured data layer, and added fault flags in place of fault history

sam-ao committed:
Merge branch 'compactUI' of https://github.com/UCSolarCarTeam/Schulich-Delta-Host-Telemetry into compactUI

sam-ao committed:
fixing conflicts with master

JordanHeinrichs committed:
Minor cleanup of whitespace.

There were a mix of 3 space and 4 space tabbing. I cleaned it up to 3 everywhere.

JordanHeinrichs committed:
Cleaned up more of the random tab sizes.

JordanHeinrichs commented:
I really like the new faults and the cleanup to the CompactUi, great job. There were inconsistencies with the tab size and whether or not spaces were used. Some of the issues were legacy and some were new. I believe 3 spaces as a tab is used within this project so I went ahead an cleaned that up.

JordanHeinrichs commented:
I would be for merging this in. I can't test it any more than just simply running it though. I think more cleanup is in order with how mpptdata is stored and done. But that goes much further than the scope of this pull request.

enochtsang commented:
Hey Jordan, looks like your pull request review has been very thorough so I'm not going to look into the code myself. I will do the blackbox testing as you have asked me to. I will merge the pull request when that is finished.

JordanHeinrichs committed:
Refactored communication layer to autoconnect.

Removed UDP connection

JordanHeinrichs committed:
Changed the label accessor functions to specify that they are returning a label.

The set prefix shouldn't be used on a getter for a reference.

JordanHeinrichs commented:
Current issues with the UI are:

* Some values under main are not populated.
* Faults lose their green background once updated.
* Connection status gray box changes.

Shown below:
![image](https://cloud.githubusercontent.com/assets/9473925/10779995/f8524856-7cfd-11e5-947f-71a1103ffb44.png)

enochtsang commented:
Thanks Jordan, i will review. I don't think the fix for the automatic connection should've been done in this pull request, but a separate pull request to master. Don't bother changing it now, but for next time.

JordanHeinrichs commented:
I agree that I did tact on a lot of extra stuff to this pull request but the changes I did to fix the connection were specific for this GUI since I used the assumption that the user will not be manually connecting. If I applied this fix to the old GUI it would break it.

enochtsang commented:
Ok, makes sense.

enochtsang commented:
When running normally, the program should be full screen. I think what we should do is add a command line option to open in windowed mode for development on a computer, but normally will build in full screen. This can be done in another pull request.

enochtsang commented:
Stage 1: Is the purpose of the pull request sound? Yes

ExtraGravity commented on CommunicationLayer/CommDeviceControl/RadioConnectionService.cpp:
```diff
@@ -30,18 +31,25 @@ bool RadioConnectionService::connectToDataSource()
 }
 
-void RadioConnectionService::disconnectFromDataSource()
+void RadioConnectionService::handleSerialDataIncoming()
```
I would disagree with the `ConnectionService` handling data incoming, it should be hooking up the `SIGNAL(readyRead())` to something that has `handleSerialDataIncoming()`. I think we should keep `CommDeviceManager` for that.

JordanHeinrichs replied:
Addressed in c6ab365004b362711b000b44dac2295c734d4dac. Renamed RadioConnectionService to be RadioCommDevice.

enochtsang commented on CommunicationLayer/CommDeviceControl/UdpConnectionService.cpp:
```diff
@@ -1,49 +0,0 @@
-#include <QUdpSocket>
```
I think this file will still be necessary, we will be connecting to the server using Udp, but it is true the logic in here doesn't make sense. I think what we should do is leave the file but turn everything into stubs.

enochtsang commented on CommunicationLayer/CommunicationContainer.cpp:
```diff
@@ -2,12 +2,9 @@
-#include "CommDeviceControl/ConnectionController.h"
-#include "CommDeviceControl/CommDeviceManager.h"
 #include "CommDeviceControl/RadioConnectionService.h"
-#include "CommDeviceControl/UdpConnectionService.h"
```
I think we still need this, we will be connecting to the server through 3G with UDP. But it is true the logic in there doesn't make sense, I think we should keep the file but turn all the functions into stubs.

enochtsang commented on CommunicationLayer/CommunicationContainer.cpp:
```diff
@@ -2,12 +2,9 @@
 #include "../DataLayer/DataContainer.h"
-#include "CommunicationContainer.h"
-#include "CommDeviceControl/ConnectionController.h"
-#include "CommDeviceControl/CommDeviceManager.h"
```
Along the same reasoning, I think this file is necessary as well, where it is in charge of connecting to all the devices.

ExtraGravity replied:
We probably still need this, we will be connecting using `UdpConnectionService` later. I can understand removing `UdpConnectionService` itself since it won't be implemented yet. But CommDeviceManager should still be connecting the serial connection.

JordanHeinrichs replied:
Here are the reasons why I don't think we need  UdpConnectionService/commDeviceManager/ConnectionController:
- These are there for the sole purpose of receiving data. For transmitting we still have the UDPForwardingService which will be modified to send to a server and not on local multicast.
- The reason they were created was to allow to dynamically switch between Radio and UDP input. The host telemetry has no reason to be listening to incoming UDP telemetry. 
- The controller needed to be there to activate and deactivate other services such as the UDP forwarding but now all those services are able to just be started on construction.
- Because the connection is extremely simplified I think it is fine to contain everything within the RadioCommDevice as all it is doing is opening up the port which is a logical construction item. Anything more than that I could justify extracting to another service.

ExtraGravity replied:
Good explanation, I'll accept that. No need for structural changes in the RadioCommDevice then.

ExtraGravity commented on DataLayer/MpptData/I_MpptData.h:
```diff
@@ -7,157 +7,218 @@ class I_MpptData : public QObject
+   virtual double mppt1CurrentIn() const = 0;
+   virtual double mppt1VoltageOut() const = 0;
+   virtual double mppt1CurrentOut() const = 0;
+   virtual double mppt1PowerIn() const = 0;
```
@JordanHeinrichs I recall you mentioning that the new MPPTs don't give both voltage and current so we can't calculate power and efficiency. Is this something that should be reflected in this UI?

JordanHeinrichs replied:
I think there could definitely be improvements to better represent that. Currently the MPPTs that don't have output current will just never emit mpptXCurrentOutReceived() but I think it should have an isValid function for each of them.

What I was going to do after this pull request is complete was to refactor this class as right now all the code is duplicated 7 times which is not good design. I was thinking that each Mppt is its own object and the MPPT data as a whole would contain 7 mppt objects. I think adding isXValid functions also should be done at that point.

ExtraGravity replied:
Sounds good. We will worry about this in a different pull request then.

enochtsang commented:
The Client Telemetry had the functionality of highlighting the highest and lowest CMU voltages, we can leave this for a different pull request, but just noting that this should be included.

ExtraGravity commented on ViewLayer/CompactUI/CompactUI.h:
```diff
@@ -0,0 +1,108 @@
+
+private:
+    Ui::CompactUI *ui;
+    QList<QLabel*> batteryCMUCellVoltageLabels_;
```
I know this was done in the Client Telemetry. but I don't think the UI file should hold this private member. This should be held in the view during runtime.

ExtraGravity commented on ViewLayer/CompactView/CompactView.cpp:
```diff
@@ -0,0 +1,491 @@
+void CompactView::motorOneFaultsReceived(MotorFaults motorFaults)
+{
+   if(motorFaults.softwareOverCurrent() && !isLabelRed(ui_.softwareOverCurrentLabel1())){
+      ui_.softwareOverCurrentLabel1().setStyleSheet("background: qlineargradient(x1: 0, y1: 0, x2: 0, y2: 1,stop: 0 rgb(176, 13, 13), stop: 1 red);");
```
I think this functionality should be abstracted into another function. Turning a faultLabel red or green.

enochtsang commented:
Overall there's a lot of string constants in the CompactView. This was poorly done by me first, but we should fix it now. Alot of the functions that edit style sheets should be abstracted into functions, because right now `setStyleSheet("some stylesheet string")` does not make it clear what it is doing.

enochtsang commented:
That's it for the changes I suggest for Stage 2 of this pull request.

JordanHeinrichs committed:
Renamed RadioConnectionService to be RadioCommDevice.

This is to better reflect what it does. There is no need for a separate connection service with the simplicity of the conection.

sam-ao committed:
fixed fault flag style sheet and connection status style sheet, added functions to clarify the changes made in the case of successful/failed connection and fault flag toggles

sam-ao commented:
![image](https://cloud.githubusercontent.com/assets/14141969/11016776/cf4842e2-8547-11e5-86c8-53f150a532d2.png)
 I added functions to change the style sheets while being clear of the function. Also fixed style sheet issues with the connection labels and the fault flags.

sam-ao committed:
did some minor restructuring

sam-ao committed:
fixing merge conflicts with master

enochtsang commented:
Starting Stage 3: Bugs and Glitches

I noticed with using the TelemTestTool, the first packet gets ignored by the Host Telemetry. Should this be fixed? I remember you commenting about it before @JordanHeinrichs

enochtsang commented:
The code in the telemTestTool for sending faults is...
```
...
   packetPayload[1] = static_cast<unsigned char>(0xFF & vehicleData_.motorOneErrorFlags);
   packetPayload[2] = static_cast<unsigned char>(0xFF & vehicleData_.motorOneLimitFlags);
   packetPayload[3] = static_cast<unsigned char>(0xFF & vehicleData_.motorTwoErrorFlags);
   packetPayload[4] = static_cast<unsigned char>(0xFF & vehicleData_.motorTwoLimitFlags);
   packetPayload[5] = static_cast<unsigned char>(0xFF & vehicleData_.bmuStatusFlagsExtended);
   packetPayload[6] = static_cast<unsigned char>(0xFF & (vehicleData_.bmuStatusFlagsExtended >> 8));
...
```

This looks like all faults should be activated but as in Sam's picture, only 3 of them get toggled, this doesn't seem like correct behaviour.

JordanHeinrichs commented:
The first packet gets ignored by design as it requires at least one packet received to establish connection. I don't see a reason to fix it since in the realistic operating condition it makes no difference. 

All faults should not be active. Look at what is assigned within the VehicleData then look at the data sheet to see what faults those numbers trigger.

sam-ao commented:
The source of the bugs with the fault flags was an issue with the test application, not the actual GUI.

enochtsang commented:
Looks good otherwise.
